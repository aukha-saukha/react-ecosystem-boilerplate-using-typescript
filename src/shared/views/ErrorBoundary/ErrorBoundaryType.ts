export interface ErrorBoundaryComponentStateInterface {
  hasError: boolean;
}

export interface ErrorInfoInterface {
  ['componentStack']: string;
}
