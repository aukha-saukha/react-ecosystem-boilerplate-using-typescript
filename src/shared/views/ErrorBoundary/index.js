import ErrorBoundary from './ErrorBoundaryComponent';

export default ErrorBoundary;
