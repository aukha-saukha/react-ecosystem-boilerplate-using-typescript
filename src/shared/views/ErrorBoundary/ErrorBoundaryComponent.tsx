import * as React from 'react';

import { ErrorBoundaryComponentStateInterface, ErrorInfoInterface } from './ErrorBoundaryType';

class ErrorBoundaryComponent extends React.Component<{}, ErrorBoundaryComponentStateInterface> {
  public readonly state: Readonly<ErrorBoundaryComponentStateInterface> = {
    hasError: false,
  };

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public static getDerivedStateFromError(_error: Error): ErrorBoundaryComponentStateInterface {
    return {
      hasError: true,
    };
  }

  public componentDidCatch(error: Error, errorInfo: ErrorInfoInterface): void {
    // TODO: Add a development envionment check for the following console.log statement,
    // and use a logger such as winston.
    // eslint-disable-next-line no-console
    console.log(error, errorInfo);
  }

  public render(): React.ReactNode {
    const { children } = this.props;
    const { hasError } = this.state;

    if (hasError) {
      return (
        <section className="display-flex flex-items-align-center flex-items-justify-content-center font-size-18 height-100vh">
          <h1>Something went wrong. Please refresh the page, and try again.</h1>
        </section>
      );
    }

    return children;
  }
}

export default ErrorBoundaryComponent;
