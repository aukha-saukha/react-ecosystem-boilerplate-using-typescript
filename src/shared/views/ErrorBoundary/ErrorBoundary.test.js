import { mount, shallow } from 'enzyme';
import React from 'react';

import ErrorBoundary from './ErrorBoundaryComponent';

function ChildWithError() {
  throw Error('Intentional error');
}

// There are console error and log messages which are typically useful, but can be suppressed
// because we intentionally introduced an error.
function swallowConsoleError(codeToRun) {
  // Copy console.error and console.log implementation in temporary variables
  const { error, log } = console;

  // Assign console.error and console.log to empty arrow functions.
  // eslint-disable-next-line no-console
  console.error = () => {};
  // eslint-disable-next-line no-console
  console.log = () => {};

  // Run the code that shows console error and log messages
  codeToRun();

  // Reimplement the orginal console.error and console.log implementation
  // eslint-disable-next-line no-console
  console.error = error;
  // eslint-disable-next-line no-console
  console.log = log;
}

describe('"Error boundary" component', () => {
  it('Renders correctly', () => {
    const wrapper = shallow(<ErrorBoundary />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Shows the correct error message.', () => {
    swallowConsoleError(() => {
      const wrapper = mount(
        <ErrorBoundary>
          <ChildWithError />
        </ErrorBoundary>
      );

      const text = wrapper.text();

      expect(text).toEqual('Something went wrong. Please refresh the page, and try again.');

      wrapper.unmount();
    });
  });

  test('Shows the correct error stack.', () => {
    swallowConsoleError(() => {
      const spy = jest.spyOn(ErrorBoundary.prototype, 'componentDidCatch');

      const wrapper = mount(
        <ErrorBoundary>
          <ChildWithError />
        </ErrorBoundary>
      );

      expect(spy).toHaveBeenCalled();

      // prettier-ignore
      expect(spy.mock.calls[0][1]).toEqual(
        {
          componentStack: `
    in ChildWithError
    in ErrorBoundaryComponent (created by WrapperComponent)
    in WrapperComponent`
        }
      );

      spy.mockClear();
      spy.mockReset();
      spy.mockRestore();
      wrapper.unmount();
    });
  });
});
