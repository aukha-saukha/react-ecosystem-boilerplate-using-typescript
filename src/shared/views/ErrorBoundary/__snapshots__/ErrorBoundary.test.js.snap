// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"Error boundary" component Renders correctly 1`] = `""`;
